design/comment-scanning.md

```diff
@@ -0,0 +1,110 @@
+customElements.define('vanilla-element2', IPromiseThisExists);
+customElements.define('vanilla-element3', SeriouslyLinterDontBeMad);
+/* polymer-lint enable: undefined-elements */
+```
+
+The polymer-lint enable/disable use-case is the immediate need here. But instead of building one-off support for this into the analyzer, we’d like to also think about how the analyzer should analyze all comment-based information so that future features don’t need to reinvent the wheel when it comes to comment-based analysis.
+
+
+## Design Overview
+
+This design groups all analysis-aware documentation into two supported formats: **Comment Directives** & **Comment Annotations**. Each supports a different general use of documentation, and together they cover most use cases for sending instructions and explicit information to the analyzer.
+
+### Comment Directive
+
+```
+/* polymer-lint disable: undefined-elements, rule-2 */
+/* polymer-lint enable: undefined-elements */
+/* @lazyImport '../some-url/some-file.html */
+<!-- @registerServiceWorkerHere -->
+```
+
+- A comment directive represents a new **Directive** feature type within a document
+- A new scanner should be written and used for each supported comment directive
+- A directive can describe one of three things:
+  - **Single Line Directive:** A feature that exists on a single line.
+  - **Source Range Directive Start:** A feature that exists on a source range within a document, starting at the directive’s source location.
+  - **Source Range Directive End:** A feature that exists on a source range within a document, ending at the directive’s source location.
+- A comment directive has a unique ID and may take arguments
+  - Because of the wide range of possible features & needs, specific syntax is lax
+  - Parsing is left up to each scanner
+- A comment directive must exist in its own comment block
+- A source range directive comment can describe several Directive features based on the arguments given
+
+
+### Comment Annotation
+
+```js
+/**
+ * @polymer
+ * @mixinFunction
+ * @memberof Polymer
+ */
+const TestMixin = function(superclass) {
+```
+
+- A comment annotation describes another, already-existing feature in a document
+- A comment annotation has a unique ID and may take arguments
+- Multiple comment annotations are allowed within a single code block
+  - Each requires a new line
+- Because they inform already-existing features, no new scanners are required
+  - The scanners for those features are responsible for their scanning
+- `@demo` & `@customElement` are two comment annotations that are already supported by the analyzer
+
+
+## Implementation Example: Polymer Lint
+
+Here is an overview of how the Linter would work using the two comment-types described above.
+
+For the purpose of this example, we’ll denote a Lint Directive as `IgnoreDirective([ruleName])` where ruleName is the name of a specific rule to disable. If no ruleName is provided, all lint rules will be disabled.
```

Comment:
I specifically ignored this use case for v1 because I'm not sure how useful it is, but happy to discuss the value in adding it.

Mainly, I couldn't think a good scenario where this would be useful. Why would a user want to enable a rule for only a single section or file? If they don't care about that rule across their entire project, why would they care in that one place? If a user was migrating to use a new rule, we should encourage them to turn it on globally and then disable individually so that the directives would be removed as the migration progressed.

FWIW, `IgnoreDirective` was pseudo code to help simplify this specific example. In the actual implementation the directive wouldn't be so coupled to ignore-only.